### PR TITLE
KC-1270: Seed MetadataCache w/ the active controller ID on MetaLogManager listener callback

### PR DESCRIPTION
**Pull request by aloknnikhil**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/metalog/MetaLogLeader.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
     private final int nodeId;
     private final long epoch;
 
-    MetaLogLeader(int nodeId, long epoch) {
+    public MetaLogLeader(int nodeId, long epoch) {
```
Comment (aloknnikhil) [inline metadata/src/main/java/org/apache/kafka/metalog/MetaLogLeader.java:29]: @cmccabe Not sure if this needs to be package private. I needed to make it public for testing in `BrokerMetadataListenerTest` which is not in the same package.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala [hunk 2]
```diff
@@ -205,6 +206,22 @@ class BrokerMetadataListener(
...
+      MetadataSnapshot(
+        snapshot.partitionStates,
+        Some(leader.nodeId), // New leader ID
+        snapshot.aliveBrokers,
+        snapshot.aliveNodes,
+        snapshot.topicIdMap,
+        snapshot.fencedBrokers,
+        snapshot.brokerEpochs))
+  }
+
```
Comment (rondagostino) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:224]: There is concurrency problem here.  `PartitionMetadataProcessor` could be processing a batch of metadata records.  When it does so, it copies the current metadata cache state,  processes all of the records, and then writes the new metadata cache state.  If we write the state here, and that occurs after `PartitionMetadataProcessor` copied it but before it wrote out the new state based on its current batch, the state we write here will get overwritten.

Note that `PartitionMetadataProcessor` never changes the controllerId value in the metadata cache, so one possibility is to make it always take the current value -- i.e. replace all its occurrences of `mgr.getMetadataSnapshot().controllerId` with metadataCache.readState().controllerId`.

File: metadata/src/main/java/org/apache/kafka/metalog/MetaLogLeader.java (hunk 1 above)
Comment (rondagostino) [reply metadata/src/main/java/org/apache/kafka/metalog/MetaLogLeader.java:29]: > Not sure if this needs to be package private. I needed to make it public for testing in BrokerMetadataListenerTest which is not in the same package.

You can make it `private[kafka]` (pretty darn close to public, of course, but not quite)

**Review by rondagostino — COMMENTED** [review]
One concurrency issue, but I outlined a quick solution.  If you go with it probably best to leave copious comments about why the code is always taking the latest controller ID instead of writing what it got when it copied the metadata cache state.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala (hunk 2 above)
Comment (rondagostino) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:224]: There's a race condition even in this "quick solution".

Comment (aloknnikhil) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:224]: Spoke with Ron offline. I think it's easier to separate the ownership for updating the MetadataSnapshot fields to not introduce these race conditions. I'll go ahead and add APIs on the MetadataCache to let the `PartitionMetadataProcessor` update all fields except for the controller ID which will be owned by the `BrokerMetadataListener`.

File: metadata/src/main/java/org/apache/kafka/metalog/MetaLogLeader.java (hunk 1 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/metalog/MetaLogLeader.java:29]: oh, `private[kafka]` doesn't exist in Java, unfortunately.  I'm OK with making it public for now.

**Review by rondagostino — APPROVED** [review]
LGTM

**Pull request merged**